File: 프로그래머스/2/118667. 두 큐 합 같게 만들기/두 큐 합 같게 만들기.cpp
#include <string>
#include <vector>
#include <numeric>
#include <queue>
#include <iostream>

using namespace std;
//한 번의 pop과 한 번의 insert를 합쳐서 작업을 1회
//어떤 방법으로도 각 큐의 원소 합을 같게 만들 수 없는 경우, -1을 return
//queue1의 길이 = queue2의 길이

int solution(vector<int> queue1, vector<int> queue2) {
    int answer = 0;
    long sum1 = accumulate(queue1.begin(), queue1.end(), 0);
    long sum2 = accumulate(queue2.begin(), queue2.end(), 0);
    long goal = (sum1 + sum2) / 2;
    queue<int> q1;
    queue<int> q2;
    for(int i = 0; i < queue1.size(); ++i)
    {
        q1.push(queue1[i]);
        q2.push(queue2[i]);
        if((sum1+sum2) % 2 == 1) return -1;
        // if(queue1[i] > goal || queue2[i] > goal) return -1;
    }


    while(true)
    {
        if(sum1 == goal) return answer;

        else if(sum1 > goal)
        {
            int temp = q1.front();
            q1.pop();
            q2.push(temp);
            sum1 -= temp;
        }
        else if(sum1 < goal)
        {
            int temp = q2.front();
            q2.pop();
            q1.push(temp);
            sum1 += temp;
        }
        answer++;

        if(answer > queue1.size() * 3) return -1;
    }
}

File: 프로그래머스/2/118667. 두 큐 합 같게 만들기/두 큐 합 같게 만들기.py
def solution(queue1, queue2):

    from collections import deque 

    sum1, sum2 = sum(queue1), sum(queue2)
    res = 0

    if sum1 == sum2: return res
    if (sum1 + sum2) % 2 == 1: return -1

    queue1 = deque(queue1)
    queue2 = deque(queue2)

    if sum1 < sum2:
        queue1, queue2 = queue2, queue1
        sum1, sum2 = sum2, sum1        

    q1 = queue1.copy()
    qSum = sum1
    l = len(queue1) * 3 - 3
    count = 0

    while sum1 != sum2:

        item = queue1.popleft()
        queue2.append(item)
        sum1 -= item
        sum2 += item
        res += 1        

        if sum2 > sum1:
            queue1, queue2 = queue2, queue1
            sum1, sum2 = sum2, sum1

        count += 1
        if count == l: break
    if sum1 != sum2: return -1

    return res